nomos-da/network/messages/src/packing.rs

```diff
@@ -0,0 +1,156 @@
+// STD
+use futures::AsyncReadExt;
+use std::io;
+// Crates
+use nomos_core::wire;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+// Internal
+use crate::Result;
+
+type LenType = u16;
+const MAX_MSG_LEN_BYTES: usize = size_of::<LenType>();
+const MAX_MSG_LEN: usize = 1 << (MAX_MSG_LEN_BYTES * 8);
+
+fn into_failed_to_serialize(error: wire::Error) -> io::Error {
+    io::Error::new(
+        io::ErrorKind::InvalidData,
+        format!("Failed to serialize message: {}", error),
+    )
+}
+
+fn into_failed_to_deserialize(error: wire::Error) -> io::Error {
+    io::Error::new(
+        io::ErrorKind::InvalidData,
+        format!("Failed to deserialize message: {}", error),
+    )
+}
+
+fn get_message_too_large_error(message_len: usize) -> io::Error {
+    io::Error::new(
+        io::ErrorKind::InvalidData,
+        format!(
+            "Message too large. Maximum size is {}. Actual size is {}",
+            MAX_MSG_LEN, message_len
+        ),
+    )
+}
+
+pub fn pack<Message>(message: &Message) -> Result<Vec<u8>>
+where
+    Message: Serialize,
+{
+    let serialized_message = wire::serialize(message).map_err(into_failed_to_serialize)?;
+
+    let data_length = serialized_message.len();
+    if data_length > MAX_MSG_LEN {
+        return Err(get_message_too_large_error(data_length));
+    }
+
+    let mut buffer = Vec::with_capacity(MAX_MSG_LEN_BYTES + data_length);
+    buffer.extend_from_slice(&(data_length as u16).to_be_bytes());
+    buffer.extend_from_slice(&serialized_message);
+    Ok(buffer)
+}
+
+async fn read_data_length<R>(reader: &mut R) -> Result<usize>
+where
+    R: AsyncReadExt + Unpin,
+{
+    let mut length_prefix = [0u8; MAX_MSG_LEN_BYTES];
+    reader.read_exact(&mut length_prefix).await?;
+    let s = u16::from_be_bytes(length_prefix) as usize;
+    Ok(s)
+}
+
+pub async fn unpack_from_reader<Message, R>(reader: &mut R) -> Result<Message>
+where
+    Message: DeserializeOwned,
+    R: AsyncReadExt + Unpin,
+{
+    let data_length = read_data_length(reader).await?;
+    let mut data = vec![0u8; data_length];
+    reader.read_exact(&mut data).await?;
+    wire::deserialize(&data).map_err(into_failed_to_deserialize)
```

Comment:
For completeness of this library, we should probably just wrap this call into an `unpack` method (?)

Reply:
I was tempted about doing that, but I felt the `unpack` would be the counterpart to `pack`. That means I'd expected it to just receive `bytes`,  not a `reader`.

This is probably the reason why the previous implementation followed this same naming scheme.

Reply:
No, I mean adding the unpack from bytes method and use it here. More or less this:

```rust
pub fn unpack<M: DeserializeOwned>(data: &[u8]) -> Result<M> {
    wire::deserialize(&data).map_err(into_failed_to_deserialize)
}
```

Reply:
Ah, yes, I thought of doing that and somehow I forgot. Thanks for pointing that out!

Reply:
The problem with that approach is it doesn't make sense conceptually. Let me elaborate:
- The `pack` function returns [size + data] as bytes.
  - To me that means I'd expected any `unpack*` function to be analogous, that is, to reconstruct from [size + data].
- The `unpack_from_reader` consumes `SizeLen` bytes, interprets that as data size, retrieves that amount of bytes as data, and then deserializes them.
- I'd expected an `unpack` function to be identical to `unpack_from_reader` but receiving `bytes` instead of a parameter. Meaning: Receiving [size + data]  as bytes.

We wouldn't be able to reuse that `unpack` function inside the `unpack_from_reader`:
- The `unpack_from_reader` would need to do something like:
  ```
  let bytes = reader.read_all()
  unpack(bytes)
  ```
  Which is not possible because we need to read an specific amount of bytes.
- Also, the `unpack` function would need to receive only [data] bytes, which feels slightly counter-intuitive to me, as previously explained.

That being said, I'm not opposed to doing it, just stating my POV :)

Reply:
I get what you mean. Ideally `pack` <=> `unpack` should be bijective, pack receives a message and returns bytes, unpack receives bytes returns a message. Maybe the correct change actually is to remove pack and have a `pack_to_writer<M, W>(message: &M, w: &mut W) -> Result<(), io::Error> where W: AsyncWriter`. Instead to fit the API in a single direction. wdyt?

Reply:
Oh, that actually makes sense and makes everything bijective. Good stuff.